README.md

```diff
@@ -100,44 +103,68 @@ t := myQueue.NewTask(&taskq.TaskOptions{
 })
 
 // Say "Hello World".
-t.Call("World")
+myQueue.Add(t.WithArgs(context.Background(), "World"))
 
 // Same using Message API.
-t.AddMessage(taskq.NewMessage("World"))
+myQueue.Add(t.WithMessage(taskq.NewMessage("World")))
 
 // Say "Hello World" with 1 hour delay.
 msg := taskq.NewMessage("World")
 msg.Delay = time.Hour
-t.AddMessage(msg)
+myQueue.Add(msg)
 
 // Say "Hello World" once.
 for i := 0; i < 100; i++ {
     msg := taskq.NewMessage("hello")
     msg.Name = "hello-world" // unique
-    t.Add(msg)
+    myQueue.Add(msg)
 }
 
 // Say "Hello World" once with 1 hour delay.
 for i := 0; i < 100; i++ {
     msg := taskq.NewMessage("hello")
     msg.Name = "hello-world"
     msg.Delay = time.Hour
-    t.Add(msg)
+    myQueue.Add(msg)
 }
 
 // Say "Hello World" once in an hour.
 for i := 0; i < 100; i++ {
-    t.CallOnce(time.Hour, "hello")
+    myQueue.Add(t.OnceWithArgs(time.Hour, "hello"))
 }
 
 // Say "Hello World" for Europe region once in an hour.
 for i := 0; i < 100; i++ {
     msg := taskq.NewMessage("hello")
     msg.OnceWithArgs(time.Hour, "europe") // set delay and autogenerate message name
-    t.Add(msg)
+    myQueue.Add(msg)
 }
 ```
 
+## Message deduplication
+If a `Message` has a `Name` then this will be used as unique identifier and messages with the same name will be deduplicated (i.e. not processed again) within a 24 hour period (or possibly longer if not evicted from local cache after that period). Where `Name` is omitted then non deduplication occurs and each message will be processed. `Task`'s `WithMessage` and `WithArgs` both produces messages with no `Name` so will not be deduplicated. `OnceWithArgs` sets a name based off a consistent hash of the arguments and a quantised period of time (i.e. 'this hour', 'today') passed to `OnceWithArgs` a `period`. This guarantees that the same function will not be called with the same arguments during `period'.
+
+## Contextual handlers
+If a task is registered with a handler that takes a Go `context.Context` as its first argument then when that handler is invoked it will be passed the same `Context` that was passed to `Consumer.Start(ctx)`. This can be used to transmit a signal to abort to all tasks being processed:
```

Comment:
One question I have about the API is that task.WithMessage takes a context that it passed to task.newMessage. Is there actually any point in this? I can't see how this context ever gets used since we serialise and deserialise the message before it ever reaches the handler. And before we handle a message we set its context to the value held by the worker instance - i.e. the worker/consumer global context.

Unless I'm missing something it seems like this per-message context could/should be removed? Perhaps there should be a unexported newMessage(ctx, args...) and an exported NewMessage(args...), otherwise you end up providing a context we know can never have an effect. If you think I'm correct on this happy to make a PR. We could still always choose to initialise the Message.Ctx to context.Background() under the hood as a nice invariant to have around.

Reply:
You are right that context is not used in practice / 99% of cases.

But there is a memqueue backend with special `SetSync(true)` option which processes message synchronously as they are added into the queue. In this case context from NewMessage is passed to the worker / handler function. The whole `SetSync` story is a hack I've created for tests and it probably can/should be removed together with context from NewMessage.

But another use case for context in NewMessage can be tracing support using opentelemetry / opencensus. I think it should be possible to extract trace id and span id from the context, save it into the Message.TraceID/SpanID, and pass it down to the worker with the message.

So I'd say that context in NewMessage is moderately useful but API is rather chatty/clumsy, e.g.

```go
msg := someLongTaskName.OnceWithArgs(ctx, time.Minute, arg1, arg2, arg3, arg4, arg5) // lots of args
_ = someLongQueueName.Add(msg)
```

But making context optional does not make it much better or even makes it worse when you actually need context

```go
msg := someLongTaskName.OnceWithArgs(time.Minute, arg1, arg2, arg3, arg4, arg5) // lots of args
msg.Ctx = ctx
_ = someLongQueueName.Add(msg)
```

We could try to play with chaining e.g.

```go
msg := someLongTaskName.NewMessage().
  WithContext(ctx).
  WithTask(someTaskName).
  WithArgs(arg1, arg2, arg3, arg4).
  OnceInPeriod(time.Minute)
_ = someLongQueueName.Add(msg)
```

Which is probably more readable...